### ISSUE TITLE
Add .data argument to map

### ISSUE THREAD

**Issue by brendanf** [ropensci/drake#719]
In various circumstances (#710, #697) it is useful to specify several precomputed values simultaneously to `map()`.  I generally make these in a data.frame/tibble, and then end up referring to them like this:
```r
meta <- tibble(all my stuff...) %>%
chain_of_dplyr_functions()

plan <- drake_plan(
    A = target(f(a, b, c, d),
       map(a = !!meta$a_,
           b = !!meta$b_,
           c = !!meta$c_,
           d = !!meta$d_,
           e = !!meta$e_,
           .id = e)))
```

Which seems needlessly verbose.  Something like
```
plan <- drake_plan(
  A = target(f(a, b, c, d),
    map(.data = meta, .id = e))
)
```
Essentially, take the supplied data frame as (part of?) the `dsl_grid()`.

I think I'm getting familiar enough with the DSL code that I may try to put something together myself.

**Comment by wlandau** [comment]
Nice! I am thinking `.data` could be a list or a data frame, and `cross()` could use it too. For the sake of clean bookkeeping, we could define a new `dsl_data()` S3 method to assign a `data` attribute in `interpret_transform()`.

https://github.com/ropensci/drake/blob/21bc2d7f998c26528806a5417dc688cf9168335b/R/api-dsl.R#L222-L228

Then we could merge the `.data` into the `new_groupings` attribute here:

https://github.com/ropensci/drake/blob/21bc2d7f998c26528806a5417dc688cf9168335b/R/api-dsl.R#L283-L301

How much time would you like to put together a PR?

**Comment by wlandau** [comment]
We may also need the `envir` passed to `transform_plan()`. Here is a sketch below, where `find_new_groupings()` becomes `explicit_new_groupings()`.

```r
new_groupings.map <- new_groupings.cross function(transform, envir) {
  attr <- attr(transform, "new_groupings")
  if (!is.null(attr)) {
    return(attr)
  }
  transform <- lang(transform)
  explicit <- explicit_new_groupings(
    transform,
    exclude = c(".data", ".id", ".tag_in", ".tag_out")
  )
  data_arg <- transform$.data
  if (!is.null(data_arg)) {
    data_arg <- eval(data_arg, envir = envir) # Get the list or data frame using the symbol in `.data`.
  }
  c(explicit, as.list(data_arg))
}


explicit_new_groupings <- function(code, exclude = character(0)) {
  list <- named(as.list(code), exclude)
  lapply(list, function(x) {
    if (is.call(x)) {
      x <- x[-1]
    }
    as.character(lapply(as.list(x), deparse))
  })
}
```

**Comment by wlandau** [comment]
Hmm... the metaprogramming may be more complicated than that because of `dsl_sym()`. I may need to work on this too.

**Comment by brendanf** [comment]
I'm not going to be able to work on it more today, I think, so by all means go for it.  :-)

I was actually thinking about what the ideal behavior of `.data` in `cross()` would be in a situation like this:

```r
x_ = c(x1, x2)
meta = tibble(y = c(y1, y2),
                      z = c(z1, z2))
drake_plan(
  A = target(f1(x),
    transform = map(x = !!x_)),
  B = target(f2(A, y, z),
       transform = cross(A, .data = meta))
)
```

Is it more useful to cross meta column-wise, or as an item?  Is it more common to want
```
f2(f1(x1), y1, z1)
f2(f1(x1), y1, z2)
f2(f1(x1), y2, z1)
f2(f1(x1), y2, z2)
f2(f1(x2), y1, z1)
f2(f1(x2), y1, z2)
f2(f1(x2), y2, z1)
f2(f1(x2), y2, z2)
```
or just
```
f2(f1(x1), y1, z1)
f2(f1(x1), y2, z2)
f2(f1(x2), y1, z1)
f2(f1(x2), y2, z2)
```
I can see use cases for both.

**Comment by wlandau** [comment]
The first option seems more consistent with the idea of a crossing.

When I first brought up `cross()`, I was thinking that the cleanest implementation of #719 would affect both `map()` and `cross()`, in which case we would get `cross(.data = your_grid)` for free. But after looking back at the code base, especially `dsl_sym()`, I am not so sure. I will have to try it out.

Here is another approach that could be cleaner. It requires focusing only on `map()`, and it allows us to require `.data` to be a data frame. First, we need a `data` S3 attribute to store the grid.

```r
interpret_transform.map <- function(transform, envir) {
  structure(
    transform,
    new_groupings = new_groupings(transform),
    data = dsl_data(transform, envir),
    deps = dsl_deps(transform)
  )
}

dsl_data <- function(...) UseMethod("dsl_data")

dsl_data.map <- function(transform, envir) {
  attr(transform, "data") %|||%
    # Using the environment should be faster than deparsing the whole
    # data frame into the language object first.
    eval(lang(transform)[[".data"]], envir = envir)
}
```

Then, we can bind all the columns of the grid together.

```r
dsl_grid.map <- function(transform, groupings) {
  tryCatch(
    cbind(as.data.frame(groupings, stringsAsFactors = FALSE), dsl_data(transform)),
    error = function(e) {
      map_grid_error(transform, groupings)
    }
  )
}
```

We should also think of a different name for `new_groupings` since `.data` has new groupings too.

**Comment by brendanf** [comment]
Here's my real life example for cross:

I tend to think of my data in blocks. In the purest cases, when I make a map I am giving a set of characteristics about a single object.  For a real example from my current project:

|TrimFile                     |Seq.Run |Plate |Well |Direction |PrimerPair |
|:----------------------------|:-------|:-----|:----|:---------|:----------|
|pb_500_002-E1f.trim.fastq.gz |pb_500  |002   |E1   |f         |its1-lr5   |
|pb_483_002-G5f.trim.fastq.gz |pb_483  |002   |G5   |f         |gits7-its4 |
|pb_483_002-G5r.trim.fastq.gz |pb_483  |002   |G5   |r         |gits7-its4 |
|pb_483_001-C1f.trim.fastq.gz |pb_483  |001   |C1   |f         |gits7-its4 |
|pb_500_001-H6f.trim.fastq.gz |pb_500  |001   |H6   |f         |its1-lr5   |
|pb_483_002-B3r.trim.fastq.gz |pb_483  |002   |B3   |r         |gits7-its4 |
|is_057_001-D5.trim.fastq.gz  |is_057  |001   |D5   |          |gits7-its4 |
|pb_483_001-C1r.trim.fastq.gz |pb_483  |001   |C1   |r         |gits7-its4 |

Each row is an input DNA sequence file, and all the other columns are metadata about that file.  If I were to use a `cross` operation, for instance to extract several different gene regions from each sequence file, then I wouldn't want to cross the values in the metadata.

I would tend to think that the most useful (but perhaps not the most expected) functionality would be to accept a `list`, and `expand.grid` on the members of that list, but allow list members which are themselves `data.frame`s.  Keep each row of the sub-`data.frame` together.

Incidentally, this is the behavior has precedent in the newest version of `tibble` and `tidyr`:
``` r
library(tibble)
library(tidyr)
test <- tibble(a = 1:3,
               b = tibble(c = 4:6,
                          d = 7:9))
tidyr::expand(test, a, b)
#> # A tibble: 9 x 3
#>       a     c     d
#>   <int> <int> <int>
#> 1     1     4     7
#> 2     1     5     8
#> 3     1     6     9
#> 4     2     4     7
#> 5     2     5     8
#> 6     2     6     9
#> 7     3     4     7
#> 8     3     5     8
#> 9     3     6     9
```

<sup>Created on 2019-02-07 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by wlandau** [comment]
Yeah, I guess these are the sorts of things that comes up once we really dig into the concept of a crossing. I think `tidyr` already handles them very well. Now that you mention them, I think that if we attempted `cross(.data = grid)` in `drake`, we would need to replicate a lot of that existing complexity, which does not seem worth the trouble. `map(.data = stuff)` should be enough regardless of the implementation strategy.

**Comment by wlandau** [comment]
Implemented in #720. I challenge you to break it :)
